src/drivers/distance_sensor/vl53lxx/vl53lxx.cpp

```diff
@@ -0,0 +1,1163 @@
+					if (want_start) {
+						start();
+					}
+
+					start();
+
+					return OK;
+				}
+
+			case SENSOR_POLLRATE_MANUAL: {
+
+					stop();
+					_measure_ticks = 0;
+					return OK;
+				}
+
+			/* adjust to a legal polling interval in Hz */
+			default: {
+					/* do we need to start internal polling? */
+					bool want_start = (_measure_ticks == 0);
+
+					/* convert hz to tick interval via microseconds */
+					unsigned ticks = USEC2TICK(1000000 / arg);
+
+					/* update interval for next measurement */
+					_measure_ticks = ticks;
+
+					/* if we need to start the poll state machine, do it */
+					if (want_start) {
+						start();
+					}
+
+					return OK;
+				}
+			}
+		}
+
+	default:
+		/* give it to the superclass */
+		return I2C::ioctl(filp, cmd, arg);
+	}
+}
+
+
+ssize_t
+VL53LXX::read(device::file_t *filp, char *buffer, size_t buflen)
+{
+	unsigned count = buflen / sizeof(struct distance_sensor_s);
+	struct distance_sensor_s *rbuf = reinterpret_cast<struct distance_sensor_s *>(buffer);
+	int ret = 0;
+
+	/* buffer must be large enough */
+	if (count < 1) {
+		return -ENOSPC;
+	}
+
+	/* if automatic measurement is enabled */
+	if (_measure_ticks > 0) {
+
+		/*
+		 * While there is space in the caller's buffer, and reports, copy them.
+		 * Note that we may be pre-empted by the workq thread while we are doing this;
+		 * we are careful to avoid racing with them.
+		 */
+		while (count--) {
+			if (_reports->get(rbuf)) {
+				ret += sizeof(*rbuf);
+				rbuf++;
+			}
+		}
+
+		/* if there was no data, warn the caller */
+		return ret ? ret : -EAGAIN;
+	}
+
+	/* manual measurement - run one conversion */
+	do {
+		_reports->flush();
+
+		/* trigger a measurement */
+		if (OK != measure()) {
+			ret = -EIO;
+			break;
+		}
+
+		while (!_collect_phase);
+
+		/* run the collection phase */
+		if (OK != collect()) {
+			ret = -EIO;
+			break;
+		}
+
+		/* state machine will have generated a report, copy it out */
+		if (_reports->get(rbuf)) {
+			ret = sizeof(*rbuf);
+		}
+
+	} while (0);
+
+	return ret;
+}
+
+
+int
+VL53LXX::readRegister(uint8_t reg_address, uint8_t &value)
+{
+	int ret;
+	uint8_t val = 0;
+
+	ret = transfer(&reg_address, sizeof(reg_address), nullptr, 0);
+
+	if (OK != ret) {
+		DEVICE_LOG("i2c::transfer returned %d", ret);
+		perf_count(_comms_errors);
+		return ret;
+	}
+
+	/* read from the sensor */
+	ret = transfer(nullptr, 0, &val, 1);
+
+	if (ret < 0) {
+		DEVICE_LOG("error reading from sensor: %d", ret);
+		perf_count(_comms_errors);
+		return ret;
+	}
+
+	ret = OK;
+	value = val;
+
+	return ret;
+
+}
+
+
+int
+VL53LXX::readRegisterMulti(uint8_t reg_address, uint8_t *value, uint8_t length)
+{
+	int ret;
+	uint8_t val[6] = {0, 0, 0, 0, 0, 0};
+	ret = transfer(&reg_address, 1, nullptr, 0);
+
+	if (OK != ret) {
+		DEVICE_LOG("i2c::transfer returned %d", ret);
+		return ret;
+	}
+
+	/* read from the sensor */
+	ret = transfer(nullptr, 0, &val[0], length);
```

Comment:
You can have a buffer overflow here if readRegisterMulti is called with `length > 6`. Can you directly use `value` instead of `val`?